### PR TITLE
Barplots and Replicates with Simulation data

### PR DESCRIPTION
**Pull request by MerktSimon**
Fixes bug #196 and resolves enhancement #210 (although here color coding is still open for discussion)

### CODE REVIEW

File: petab/visualize/plotting_config.py [hunk 1]
```diff
@@ -72,42 +72,41 @@ def plot_lowlevel(vis_spec: pd.DataFrame,
         # add xOffset
         conditions = conditions + vis_spec.xOffset[i_visu_spec]
 
+        # TODO sort mean and sd/sem by x values (as for simulatedData below)
```
Comment (dweindl) [inline petab/visualize/plotting_config.py]: Codacy Issue found: [TODO sort mean and sd/sem by x values (as for simulatedData below)](https://app.codacy.com/manual/dweindl/PEtab/pullRequest?prid=4827411)

File: petab/visualize/plot_data_and_simulation.py [hunk 2]
```diff
@@ -122,12 +143,21 @@ def plot_data_and_simulation(
...
+            sns.despine()
+            plt.xticks(rotation=45, ha="right")
+            plt.tight_layout()
+            plt.savefig(f'{subplot_file_path}/{var_plot_id}.png')
+            plt.close()
 
     # finalize figure
-    fig.tight_layout()
-    sns.despine()
+    if plots_to_file:
```
Comment (dweindl) [inline petab/visualize/plot_data_and_simulation.py]: Codacy Issue found: [Unnecessary "else" after "return"](https://app.codacy.com/manual/dweindl/PEtab/pullRequest?prid=4827411)

File: petab/visualize/plot_data_and_simulation.py [hunk 3]
```diff
@@ -8,6 +8,8 @@
 import petab
 
 from typing import Union, Optional, List
+import matplotlib.pyplot as plt
```
Comment (dweindl) [inline petab/visualize/plot_data_and_simulation.py]: Codacy Issue found: [Unable to import 'matplotlib.pyplot'](https://app.codacy.com/manual/dweindl/PEtab/pullRequest?prid=4827411)

File: petab/visualize/plot_data_and_simulation.py [hunk 4]
```diff
@@ -104,14 +114,25 @@ def plot_data_and_simulation(
     # get unique plotIDs
     uni_plot_ids, _ = np.unique(vis_spec.plotId, return_index=True)
 
-    fig, ax, num_row, num_col = create_figure(uni_plot_ids)
+    # Switch saving plots to file on or get axes
+    plots_to_file = False
+    if subplot_file_path != '':
+        plots_to_file = True
+    else:
+        fig, ax, num_row, num_col = create_figure(uni_plot_ids, plots_to_file)
```
Comment (dweindl) [inline petab/visualize/plot_data_and_simulation.py]: Codacy Issue found: [Unused variable 'num_row'](https://app.codacy.com/manual/dweindl/PEtab/pullRequest?prid=4827411)

File: petab/visualize/plot_data_and_simulation.py [hunk 5]
```diff
@@ -104,14 +114,25 @@ def plot_data_and_simulation(
...
 
     # loop over unique plotIds
     for i_plot_id, var_plot_id in enumerate(uni_plot_ids):
 
-        # setting axis indices
-        i_row = int(np.ceil((i_plot_id + 1) / num_col)) - 1
-        i_col = int(((i_plot_id + 1) - i_row * num_col)) - 1
+        if plots_to_file:
+            fig, ax, _, num_col = create_figure(uni_plot_ids,
+                                                      plots_to_file)
```
Comment (dweindl) [inline petab/visualize/plot_data_and_simulation.py]: Codacy Issue found: [Wrong continued indentation (remove 6 spaces).](https://app.codacy.com/manual/dweindl/PEtab/pullRequest?prid=4827411)

File: petab/visualize/plot_data_and_simulation.py (hunk 3 above)
Comment (dweindl) [inline petab/visualize/plot_data_and_simulation.py]: Codacy Issue found: [Unable to import 'matplotlib.pyplot'](https://app.codacy.com/manual/dweindl/PEtab/pullRequest?prid=4827411)

File: petab/visualize/plotting_config.py (hunk 1 above)
Comment (dweindl) [inline petab/visualize/plotting_config.py]: Codacy Issue found: [TODO sort mean and sd/sem by x values (as for simulatedData below)](https://app.codacy.com/manual/dweindl/PEtab/pullRequest?prid=4827411)

File: petab/visualize/plot_data_and_simulation.py (hunk 4 above)
Comment (dweindl) [inline petab/visualize/plot_data_and_simulation.py]: Codacy Issue found: [Unused variable 'num_row'](https://app.codacy.com/manual/dweindl/PEtab/pullRequest?prid=4827411)

File: petab/visualize/plot_data_and_simulation.py (hunk 2 above)
Comment (dweindl) [inline petab/visualize/plot_data_and_simulation.py]: Codacy Issue found: [Unnecessary "else" after "return"](https://app.codacy.com/manual/dweindl/PEtab/pullRequest?prid=4827411)

File: petab/visualize/plotting_config.py [hunk 6]
```diff
@@ -72,8 +72,9 @@ def plot_lowlevel(vis_spec: pd.DataFrame,
         # add xOffset
         conditions = conditions + vis_spec.xOffset[i_visu_spec]
 
-        # TODO sort mean and sd/sem by x values (as for simulatedData below)
+        # TO DO sort mean and sd/sem by x values (as for simulatedData below)
```
Comment (dweindl) [inline petab/visualize/plotting_config.py]: Please keep as "TODO", that is what most tools recognize. "TO DO" is just regular text without special meaning.

Comment (dweindl) [reply petab/visualize/plotting_config.py]: + for multiline TODOs, add 2 blanks in the beginning of the continuation lines, then it is recognized as one single issue

**Comment by MerktSimon** [comment]
since I commited with wrong automatic issue closure keyword, again:  This pull request also closes #213 in both ways

File: petab/visualize/plotting_config.py [hunk 7]
```diff
@@ -72,42 +72,42 @@ def plot_lowlevel(vis_spec: pd.DataFrame,
         # add xOffset
         conditions = conditions + vis_spec.xOffset[i_visu_spec]
 
+        # TODO sort mean and sd/sem by x values (as for simulatedData below)
```
Comment (dweindl) [inline petab/visualize/plotting_config.py]: Codacy Issue found: [TODO sort mean and sd/sem by x values (as for simulatedData below)](https://app.codacy.com/manual/dweindl/PEtab/pullRequest?prid=4827411)

**Pull request merged**